### PR TITLE
Allow dropdowns to display un-contained values

### LINKED ISSUE #1945: Allow dropdown to handle invalid values
A `Dropdown` will currently fault when the underlying `Bindable` is updated to a value not present in the dropdown.

Here are two ideas to circumvent this:
- Add the "invalid" value as a temporary `DropdownMenuItem` and delete it once the selection is changed.
This would allow to display the selected value temporarily and allow the user to keep it (which may or may not be desirable depending on the situation)
- Restore the "default" value and update the `Bindable` accordingly.
Could be either a fallback to the first item in the list or an optional delegate to retrieve a valid value from an invalid one (more flexible).
This might result in an infinite feedback loop of `Bindable` updates if used improperly.

Causes ppy/osu#3474

### PR DESCRIPTION
**Pull request by AtomCrafty**
Closes #1945
Needs to be merged together with ppy/osu#3570

### CODE REVIEW

**Comment by AtomCrafty** [comment]
Untested and doesn't contain the restructuring proposed by @peppy, this is just a proof of concept at this point.

**Comment by AtomCrafty** [comment]
Changes so far:
- `Items` was renamed to `Entries` since it is still required by `ToolbarAssemblySection`
- New `Items` property of type `IEnumerable<T>`
- New virtual method `string GenerateItemText(T)`
- New method `void AddDropdownItem(T)`

Changes *not* made:
- Delete `Entries` completely (I'd vote for keeping it)
- Delete `void AddDropdownItem(string, T)` completely.
It is used internally anyway, so it might as well be public.
Maybe merge both methods to `void AddDropdownItem(T, [string])` where a `null` string is coalesced to `GenerateItemText(value)`?

**Comment by AtomCrafty** [comment]
Closing this for now since I don't see this model going anywhere.
Most dropdowns in lazer specify a "Default" item ("Default resolution", "Default audio device", etc.), which is impossible with the proposed changes.
Allowing it would require subclassing both `Dropdown` and creating a second version of `SettingsDropdown`, to allow specifying custom item labels, aka the exact feature this would remove.

Having tried out several implementations for this, I can state that each one of them would result in a very convoluted mess of classes or make the existing ones obsolete because everyone would need to use the new ones.

dropdown

If someone could show me an example of how this inheritance model would work in a real life scenario (for example [here](https://github.com/ppy/osu/blob/master/osu.Game/Overlays/Settings/Sections/Audio/AudioDevicesSettings.cs#L51)), I'd be willing to pick it back up, but right now I'm unsure how to move forward.

**Comment by AtomCrafty** [comment]
Meh... that was a bad example tbh.

Take a look at the [resolution](https://github.com/ppy/osu/blob/master/osu.Game/Overlays/Settings/Sections/Graphics/LayoutSettings.cs#L120) or [skin](https://github.com/ppy/osu/blob/master/osu.Game/Overlays/Settings/Sections/SkinSection.cs#L65) dropdowns.
The latter has the additional problem that a new intermediate `Bindable` and a ton of converting code is needed to register the config bindable of type `int` to a dropdown of type `SkinInfo`, because the dropdown could no longer directly select integers as values, since they would be displayed incorrectly.

**Comment by peppy** [comment]
Both of these should effortlessly convert. The skin one is fixed by using SkinInfo instead of the ID (and giving SkinInfo a ToString or Description field. The resolution one is fixed by making a class to hold resolution information (i'd argue it should be called something else as it will hold refresh rate too) and then subclassing for the default, with the special values of 9999.

**Comment by AtomCrafty** [comment]
Both thode solutions don't address the incompatibility of bindable types. The SizeFullscreen bindable requires a Size object and Skin requires an int. Two way conversion logic would be needed in several places.

**Comment by peppy** [comment]
That's fine for now – this is intended. Eventually, for example, configuration should be able to handle the generic T entities to actually save them (to database, not file, at that).

**Comment by AtomCrafty** [comment]
Ready for review.

**Comment by smoogipoo** [comment]
@peppy this is fine to merge from my side; can you check the osu! PR and see if you're fine with it?

**Comment by peppy** [comment]
Let me check this PR for correct behaviour.

File: osu.Framework/Graphics/UserInterface/Dropdown.cs [hunk 1]
```diff
@@ -57,12 +57,18 @@ public abstract class Dropdown<T> : FillFlowContainer, IHasCurrentValue<T>
...
+        /// <param name="value">Value selected by the menu item.</param>
+        public void AddDropdownItem(T value) => AddDropdownItem(GenerateItemText(value), value);
+
         /// <summary>
         /// Add a menu item directly.
         /// </summary>
         /// <param name="text">Text to display on the menu item.</param>
         /// <param name="value">Value selected by the menu item.</param>
-        public void AddDropdownItem(string text, T value)
+        protected void AddDropdownItem(string text, T value)
```
Comment (smoogipoo) [inline osu.Framework/Graphics/UserInterface/Dropdown.cs]: It's not a super big issue but this really doesn't need to be `protected`. All cases can be solved by using a more complex object as the items of the dropdown. E.g.:

```
struct AssemblyDescriptor
{
    public Assembly Assembly;
    public string Name;
}
```
(in the current use-case)

**Pull request merged**